drivers/flash/flash_esp32.c

```diff
@@ -76,26 +76,67 @@ static inline void flash_esp32_sem_give(const struct device *dev)
 #include <stdint.h>
 #include <string.h>
 
+#ifdef CONFIG_MCUBOOT
+#define READ_BUFFER_SIZE 32
+static bool flash_esp32_is_aligned(off_t address, void *buffer, size_t length)
+{
+	/* check if address, buffer pointer, and length are 4-byte aligned */
+	return ((address & 3) == 0) && (((uintptr_t)buffer & 3) == 0) && ((length & 3) == 0);
+}
+#endif
+
 static int flash_esp32_read(const struct device *dev, off_t address, void *buffer, size_t length)
 {
 	int ret = 0;
 
 #ifdef CONFIG_MCUBOOT
-	/* ensure everything is 4-byte aligned */
-	size_t aligned_length = ROUND_UP(length, 4);
-	off_t aligned_address = ROUND_DOWN(address, 4);
-	size_t address_offset = address - aligned_address;
-	uint32_t temp_buf[aligned_length / 4];
-
-	if (!esp_flash_encryption_enabled()) {
-		ret = esp_rom_flash_read(aligned_address, temp_buf, aligned_length, false);
-	} else {
-		ret = esp_rom_flash_read(aligned_address, temp_buf, aligned_length, true);
+	uint8_t *dest_ptr = (uint8_t *)buffer;
+	size_t remaining = length;
+	size_t copy_size = 0;
+	size_t aligned_size = 0;
+	bool allow_decrypt = esp_flash_encryption_enabled();
+
+	if (flash_esp32_is_aligned(address, buffer, length)) {
+		ret = esp_rom_flash_read(address, buffer, length, allow_decrypt);
+		return -EIO;
```

Comment:
shouldnt it return `ret` instead of `-EIO`  ?

Reply:
Please, check again.